objects/drawin.c

```diff
@@ -584,16 +590,48 @@ luaA_drawin_get_height(lua_State *L, drawin_t *drawin)
  * \param drawin The drawin object.
  * \return The number of elements pushed on stack.
  */
-static int
-luaA_drawin_set_ontop(lua_State *L, drawin_t *drawin)
+void
+drawin_set_ontop(lua_State *L, drawin_t *drawin, bool b)
 {
-    bool b = luaA_checkboolean(L, -1);
     if(b != drawin->ontop)
     {
+        if(b)
+            drawin_set_desktop(L, drawin, false);
         drawin->ontop = b;
         stack_windows();
         luaA_object_emit_signal(L, -3, "property::ontop", 0);
     }
+}
+
+/** Set the drawin on desktop status.
+ * \param L The Lua VM state.
+ * \param drawin The drawin object.
+ * \return The number of elements pushed on stack.
+ */
+void
+drawin_set_desktop(lua_State *L, drawin_t *drawin, bool b)
+{
+    if(b != drawin->desktop)
+    {
+        if(b)
+            drawin_set_ontop(L, drawin, false);
```

Comment:
This doesn't necessarily work. Lua could react to `property::ontop` and just set `d.ontop = true` again. Thus, you are not guaranteed that `drawin->ontop = false` after this call. I fear one would have to defer all signal emission until the end of this function to work-around this (not that the result would then look good from Lua's point of view, but at least that should ensure that at most one of these properties can be `true` at the same time).

I think these "multiple exclusive states"-thing also came up in `client.c`, but I don't know how it was handled there. I fear "not nicely"... :-(

Somehow this feels like it should be an enum, but that would break the magic behind `LUA_OBJECT_EXPORT_PROPERTY`. I don't have any good suggestion, sorry.